Transitions

claireh93 opened the pull request:
Quite a lot of changes in here, particularly:

- Bring `Ecosystem` and `EpiSystem` into a single type so that all simulations run from the same base.
- Do the same for `SpeciesList` and `EpiList`. 
- Update creation of `TransitionList` so that they are initially empty, and subsequently appended to using `addtransitions!`
- Add some new epi transitions to run on full SEI3HRD framework.
- Keep all old code functionality, but rename, i.e. `biodiversity_simulate!` and `epi_simulate` for the two code bases respectively.

richardreeve commented:
@claireh93 Should we merge main back into dev before you do this, just so they don't diverge too far?

claireh93 commented:
Good idea!

richardreeve commented:
Once you've released 0.1 and it is on the registry, do you want to remerge main into this before we merge it into dev, so dev is fully up to date, or do you want to do it in two stages?

claireh93 commented:
I think maybe do it in two stages, this branch is already quite cluttered up with changes.

richardreeve commented:
What's going on with this tif? https://github.com/boydorr/EcoSISTEM.jl/blob/f93e3a6cbafa84c2a347717df8cd41a243a906f3/examples/Africa_run.jl#L11-L13

richardreeve commented:
I presume you can get rid of this? https://github.com/boydorr/EcoSISTEM.jl/blob/f93e3a6cbafa84c2a347717df8cd41a243a906f3/src/Biodiversity/MPIEcosystem.jl#L3-L7

richardreeve commented:
This should be `biodiversity_simulate_record!()` https://github.com/boydorr/EcoSISTEM.jl/blob/f93e3a6cbafa84c2a347717df8cd41a243a906f3/src/Biodiversity/Helper.jl#L40-L42

richardreeve commented:
Likewise this... This should be `biodiversity_simulate_record!()` https://github.com/boydorr/EcoSISTEM.jl/blob/f93e3a6cbafa84c2a347717df8cd41a243a906f3/src/Biodiversity/Helper.jl#L8-L10

richardreeve commented:
I'm also not sure why some of these `biodiversity_simulate*` `epi_simulate*` and `simulate*` functions operate on `AbstractEcosystem`s and some on `Ecosystem`s?

claireh93 commented:
> What's going on with this tif?
> 
> https://github.com/boydorr/EcoSISTEM.jl/blob/f93e3a6cbafa84c2a347717df8cd41a243a906f3/examples/Africa_run.jl#L11-L13

Ah yes, I forgot about that - it's to give me the outline of the continent. As it's only 5kb I could add it into the examples folder?

claireh93 commented:
> I'm also not sure why some of these `biodiversity_simulate*` `epi_simulate*` and `simulate*` functions operate on `AbstractEcosystem`s and some on `Ecosystem`s?

I think this is just an oversight from when we had both `Ecosystem` and `EpiSystem` - I have changed them all to run on `Ecosystem`s now, because presumably if someone was implementing their own ecosystem type, they would need to design their own simulate functions anyway.

richardreeve reviewed (commented):
I got part way through this and ran out of time. A few other comments...

richardreeve commented:
Yes, adding it to the examples folder should be fine.

claireh93 commented on src/Biodiversity/Movement.jl:
```diff
@@ -109,3 +109,10 @@ getkernels(m::NoMovement) = m.kernels
+
+getdispersaldist(m::BirthOnlyMovement, sp::Int64) = m.kernels[sp].dist
+getdispersalvar(m::BirthOnlyMovement, sp::Int64) = (m.kernels[sp].dist)^2 * pi / 4
+getdispersaldist(m::AlwaysMovement, sp::Int64) = m.kernels[sp].dist
+getdispersalvar(m::AlwaysMovement, sp::Int64) = (m.kernels[sp].dist)^2 * pi / 4
+getdispersaldist(m::NoMovement, sp::Int64) = m.kernels[sp].dist
+getdispersalvar(m::NoMovement, sp::Int64) = (m.kernels[sp].dist)^2 * pi / 4
```
They are tested and I think it is helpful to have a way to see for a given species how far we expect them to move? They are called by `getdispersal*(eco, sp)` in `TransitionSystemHelperFuns.jl`- so you can ask the ecosystem directly

richardreeve replied:
Are you sure they are tested? The reason I made that comment is that in `Files changed`, those lines were flagged as added but not tested... also, I'm not sure what movement means in a `NoMovement` scenario!

claireh93 replied:
Oh yes, have added those in now and altered NoMovement so it tells you that there is no movement instigated in the system.

claireh93 commented on examples/env_transitions.jl:
```diff
@@ -69,21 +71,42 @@ category_map = (
 plot_epidynamics(epi, abuns, category_map = category_map)
 
-epi = EpiSystem(epilist, epienv, rel)
+transitions = Simulation.create_transition_list_SEIR(epilist, epienv)
```
Currently they aren't being tested, originally because `Scotland_run.jl` etc. were quite lengthy. We could reconsider this now though?

claireh93 commented:
> > I'm also not sure why some of these `biodiversity_simulate*` `epi_simulate*` and `simulate*` functions operate on `AbstractEcosystem`s and some on `Ecosystem`s?
> 
> I think this is just an oversight from when we had both `Ecosystem` and `EpiSystem` - I have changed them all to run on `Ecosystem`s now, because presumably if someone was implementing their own ecosystem type, they would need to design their own simulate functions anyway.

Oops it turns out it's so `MPIEcosystem`s can be run - I have changed them back to `AbstractEcosystem` for now, and we can think about this further once we start reimplementing the MPI.

richardreeve reviewed (changes requested):
It's hard to assess properly because the changes are so substantial, I think I'll have to play around with it once it's merged, but I think it's basically good to go. There are a few places where the docs need fleshing out a bit more or correcting that I've spotted, but I don't think I've noticed anything significant. Good work!

claireh93 commented on src/Biodiversity/Traitfuns.jl:
```diff
@@ -79,3 +79,15 @@ Function to extract the trait relationship of all species in the ecosystem.
+    return _gettraits(eco.spplist)
+end
+
+function _gettraits(sppl::SpeciesList)
+    return sppl.species.traits
+end
+function _gettraits(sppl::SpeciesList{A, B, C}) where {A <: HumanTypes, B <: VirusTypes, C}
+    return sppl.pathogens.traits
+end
```
Yes this is a bit tricky currently - for the COVID model (i.e. `HumanTypes` + `VirusTypes`), I need to access virus traits because it is the virus that is directly impacted by the environment. But for the biodiversity (and presumably other models), I need the species themselves to respond.

richardreeve replied:
I think this may be the only thing I can see that needs to be resolved before we merge. It's not really a good solution, and there's a danger we'll be stuck with it for a while if we accept it now.

richardreeve commented on src/Transitions/Transitions.jl:
```diff
@@ -1,35 +1,157 @@
+    AbstractPlaceTransition <: AbstractTransition
+
+Abstract type for place transitions. Place transitions occur across multiple grid
+squares for a single type, e.g. dispersal of types across the landscape.
```
```suggestion
squares for a single type, e.g. dispersal of seeds across the landscape.
```
